docs/core/extensions/artificial-intelligence.md

```diff
@@ -0,0 +1,650 @@
+
+> [!IMPORTANT]
+> For more usage examples and real-world scenarios, see [AI for .NET developers](../../ai/index.yml).
+
+**In this section**
+
+- [The `IChatClient` interface](#the-ichatclient-interface)
+  - [Request chat completion](#request-chat-completion)
+  - [Request chat completion with streaming](#request-chat-completion-with-streaming)
+  - [Tool calling](#tool-calling)
+  - [Cache responses](#cache-responses)
+  - [Use telemetry](#use-telemetry)
+  - [Provide options](#provide-options)
+  - [Functionality pipelines](#functionality-pipelines)
+  - [Custom `IChatClient` middleware](#custom-ichatclient-middleware)
+  - [Dependency injection](#dependency-injection)
+- [The `IEmbeddingGenerator` interface](#the-iembeddinggenerator-interface)
+  - [Sample implementation](#sample-implementation)
+  - [Create embeddings](#create-embeddings)
+  - [Custom `IEmbeddingGenerator` middleware](#custom-iembeddinggenerator-middleware)
+
+### The `IChatClient` interface
+
+The following sample implements `IChatClient` to show the general structure.
+
+:::code language="csharp" source="snippets/ai/ConsoleAI/SampleChatClient.cs":::
+
+You can find other concrete implementations of `IChatClient` in the following NuGet packages:
+
+- [📦 Microsoft.Extensions.AI.AzureAIInference](https://www.nuget.org/packages/Microsoft.Extensions.AI.AzureAIInference): Implementation backed by [Azure AI Model Inference API](/azure/ai-studio/reference/reference-model-inference-api).
+- [📦 Microsoft.Extensions.AI.Ollama](https://www.nuget.org/packages/Microsoft.Extensions.AI.Ollama): Implementation backed by [Ollama](https://ollama.com/).
+- [📦 Microsoft.Extensions.AI.OpenAI](https://www.nuget.org/packages/Microsoft.Extensions.AI.OpenAI): Implementation backed by either [OpenAI](https://openai.com/) or OpenAI-compatible endpoints (such as [Azure OpenAI](https://azure.microsoft.com/products/ai-services/openai-service)).
+
+#### Request chat completion
+
+To request a completion, call the <xref:Microsoft.Extensions.AI.IChatClient.CompleteAsync*?displayProperty=nameWithType> method. The request is composed of one or more messages, each of which is composed of one or more pieces of content. Accelerator methods exist to simplify common cases, such as constructing a request for a single piece of text content.
+
+:::code language="csharp" source="snippets/ai/ConsoleAI/Program.cs":::
+
+The core `IChatClient.CompleteAsync` method accepts a list of messages. This list represents the history of all messages that are part of the conversation.
+
+```csharp
+using Microsoft.Extensions.AI;
+
+IChatClient client = new SampleChatClient(
+    new Uri("http://coolsite.ai"), "my-custom-model");
+
+Console.WriteLine(await client.CompleteAsync(
+[
+    new(ChatRole.System, "You are a helpful AI assistant"),
+    new(ChatRole.User, "What is AI?"),
+]));
+```
```

Comment:
I'd feel a lot better if all these snippets were in a separate code file so our CI build validates them anytime they change.